### DISCUSSION TITLE
Better conflict marker

### DISCUSSION THREAD

**Discussion by black7375** [arxanas/git-branchless#923] [Ideas]
One of the things I want to improve by using git is conflict marker.

### Text

[Jujutsu's conflict marker](https://github.com/martinvonz/jj/blob/main/docs/conflicts.md#conflict-markers) looks pretty good to me.
```
  <<<<<<<
  %%%%%%% # Start diff
   apple
  -grape
  +grapefruit
   orange
  +++++++ # Start Snapshot
  APPLE
  GRAPE
  ORANGE
  >>>>>>>
```

One concern is the syntax highlighting of IDEs and editors.
And if you have less changes, it may be better to show two continuous diffs.
I think the consistent way is to use [Unified Format](https://www.gnu.org/software/diffutils/manual/html_node/Unified-Format.html).
```
Diff and Snapshot
  <<<<<<<
  @@ -1,4 +1,4 @@ Start diff base-ours
   apple
  -grape
  +grapefruit
   orange
  ======= # Start Snapshot
  APPLE
  GRAPE
  ORANGE
  >>>>>>>

Continuous diff style
  <<<<<<<
  @@ -1,4 +1,4 @@ Start diff base-ours
   apple
  -grape
  +grapefruit
   orange
  @@ -1,4 +1,5 @@ Start diff base-theirs
  -apple
  +banana
   grape
   orange
  +tomato
  >>>>>>>
```

Syntax highlighting problem may still not be solved, but I think we can lower the overhead if someonw need to implement highlighting because it is based on existing tools.

### Binary

The current behavior of git when [binary file conflict is very embarrassing](https://stackoverflow.com/questions/278081/resolving-a-git-conflict-with-binary-files).
```
## Binary conflict
warning: Cannot merge binary files: <path>
Auto-merging <path>
CONFLICT (content): Merge conflict in <path>
Automatic merge failed; fix conflicts and then commit the result.

## Then
git checkout --ours -- <path>
git checkout --theirs -- <path>
```

I solve it with a [very simple program](https://github.com/black7375/binary-conflict) that creates additional `base` and `theirs` files.
![image](https://user-images.githubusercontent.com/25581533/235336209-b818da65-9a7c-4d80-af6f-c468c856e216.png)

### Totally

We can also combine the above styles to handle a little smarter.
- Diff and Snapshot: When the change is large, it is shown as a snapshot.
- Continuous diff: If the change is trivial, it is shown as continuous diff.
- File: For binary files, create `base` and `theirs` files.

In summary, there are 4 styles that can be added.
- Diff and Snapshot
- Continuous diff
- File
- Auto

### CLI interface
Currently, [`git merge`](https://git-scm.com/docs/git-merge) cannot be used as [an argument like `git checkout --conflict=diff3 <path>`](https://git-scm.com/docs/git-checkout), and you need to use [merge-settings](https://git-scm.com/docs/merge-config).

I want to see the `--conflict=` arguments in `git merge`.
```shell
git merge featA
git merge --abort
git merge featA --conflict=diff3 # Try other conflict style
```

**Comment by arxanas** [comment]
I agree that the conflict marker situation in Git should be improved, but this isn't directly customizable in Git (you can't override `git merge`, for example), so maybe you should make your proposal in the Jujutsu discussion board instead?

**Reply by black7375** [reply]
When I came up with this idea, I naively thought, "Well, we can solve it with `git hook` and `merge driber setting`".
I just looked for it and it didn't seem easy to override the default setting just by setting `.git`.

I'll try to find out if it's possible over the weekend, but it looks good to open the issue to Jujutsu first.
